### PR TITLE
Throttle operations early not to block server

### PR DESCRIPTION
**Pull request by tezc**
Throttle operations early not to block server

Introduced throttling in https://github.com/RedisLabs/raft/pull/106. If execution of operations take as much as `request-timeout`, it means this server will not generate responses on time for the existing requests. This is causing extra elections. I've changed deadline as `request_timeout / 2`, giving server some room for other operations and enough time to send responses.

### CODE REVIEW

File: src/raft_server.c [hunk 1]
```diff
@@ -2230,8 +2230,11 @@ int raft_pending_operations(raft_server_t *me)
 
 int raft_exec_operations(raft_server_t *me)
 {
+    /* Operations should take less than `request-timeout`. If we block here more
+     * than request-timeout, it means this server won't generate responses on
+     * time for the existing requests. */
+    me->exec_deadline = raft_time_millis(me) + (me->request_timeout / 2);
```
Comment (sjpotter) [inline src/raft_server.c:2236]: what is raft_time_millis i don't see a description on it in raft code

Comment (sjpotter) [reply src/raft_server.c:2236]: no, was just wondering if it was some sort of calculation.  I'm still wondering if this is good enough.  

should the deadline be measured since the last command being sent?  i.e. is it possible that a half request_timeout might still be too much?

Comment (tezc) [reply src/raft_server.c:2236]: `request_timeout` is heartbeat interval, so, I think it will always be greater than ~100-200 milliseconds. So, let's assume it is 200 milliseconds. We spend half of it for executing operations and leaving other half for other things. Nothing else in the system should take 100 milliseconds. Normally, an event loop iteration should take ~10-100 **microseconds** at most. So, if there is something else in the system that takes 100 milliseconds(e.g VM freeze, threads pending to be scheduled, fsync of a file), it must be exceptional, in that case, failure is acceptable I guess. 

More importantly, if there is something in the system that can take 100 milliseconds, it can also take 200 milliseconds(more than request timeout) as it is something exceptional. This is always the problem, user should configure request timeout according to this. I feel like half of request timeout is much more than enough, it should be :) Do you see any other thing that can cause a failure here?

Comment (tezc) [reply src/raft_server.c:2236]: Throttling is just a best effort thing. It's fair to assume nothing else in the system will take something like 100 ms or 200 ms on each iteration. We have to make some assumptions because you'll never know accurate timestamps. For example:

- request timeout is 200 milliseconds.
- nodeA sends a request, timestamp is 1000 when nodeA writes it to the network.
- nodeB is busy with existing tasks, it processes existing requests.
- nodeB reads the request from the network, timestamp is 1100.
- nodeB thinks it should reply until 1100 + 200 = 1300 which is wrong. It will expire at 1200. 

So, it doesn't make sense trying to be accurate. We just need to make some assumptions. For example, currently, we have some assumptions, we think nothing in the system will take longer than election timeout or request timeout.

In summary, if there is an occasional event/events in the system that takes some time up to request timeout and if we have lots of operations to be executed at the same time, we can fail to reply on time. It's okay to fail in this case, this should be super rare and we retry failed requests anyway. If this scenario is happening all the time, either application should fix that source of latency (I don't think this is normal) or application should increase request timeout. 

We can lower the execution time, make it something like `request_timeout / 4` but in any case, we should make an assumption. It's fair to assume a single "event loop / application logic" iteration is pretty fast (normally, should be something like **a few microseconds** to **100 microseconds** at most).

**Review by sjpotter — APPROVED** [review]
